src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -62,7 +63,7 @@ public void Set(IEnumerable<IAnalysisIssueVisualization> issueVisualizations)
                 throw new ArgumentNullException(nameof(issueVisualizations));
             }
 
-            taintVulnerabilities.Clear();
+            taintVulnerabilities.RemoveAll();
```

Comment:
There is a problem with the Clear method:
It raises NotifyCollectionChanged with `e.OldItems = null`, and `e.Action = NotifyCollectionChangedAction.Reset`. 
It would be fine for the view model (as we discussed, it might not need the observable to begin with), but `IssueStoreObserver` relies on it too: https://github.com/SonarSource/sonarlint-visualstudio/blob/branch-taint/src/IssueViz.Security/IssueStoreObserver.cs#L116

If we don't supply the removed items to `IssueStoreObserver`, the tags don't get removed (see #2064)

Reply:
Removing one item at a time and triggering a chain of further update events feels wrong.

Taking a step back, the data structures and event forwarding just feel too complicated. When reviewing your other PR earlier, I was trying to remember what the point of the `IssueStoreObserver` was. It looks like an adapter that implements `IIssueLocationStore` over an `ObservableCollection`.

If an `ObservableCollection` isn`t the right data structure for the `TaintStore`, we shouldn`t need to care about `IssueStoreObserver`. Wouldn`t it be simpler to just to pick the right data structure, implement `IIssueLocationStore` directly on `TaintStore` and forget about `IssueStoreObserver`?

If I remember correctly, `IssueStoreObserver` was introduced to avoid duplicating code between the hotspots store and the taint store. However, the code is starting to feel over-complicated.

The update scenarios of the hotspot and taint stores are different; hotspots can be added incrementally as the user chooses to "open in IDE", whereas the taint list is only bulk-updated. However, we can't do bulk notifications with the current implementation; we have to raise one notification for each item added as part of a bulk operation, and now one notification for each item removed.

I think should have an offline discussion about this tomorrow to discuss whether the current implementation is still the right one, and if not when/how to change.

Reply:
I think `IssueStoreObserver` is an important component: it has a sizable chunk of code and that code already had bugs in it (#1958). It's identical for hotspots and taint and I don't think it's right to duplicate it. Also, we discussed that we could have a list of all the hotspots from the server (just like taint), in which case the hotspots could also be bulk-updated.

Perhaps we need to look into writing our own `ObservableCollection`, one that properly implements Clear() and bulk-notifications. Maybe there's even a package for it already.

Reply:
Hey look at [that](https://github.com/dotnet/runtime/issues/18087#issuecomment-359197102), we might get it if we're lucky 😆

Reply:
FYI: https://github.com/dotnet/runtime/issues/18087#issuecomment-359197102